### PR TITLE
Rebuild group automatically in dynamic graph distributed

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

New features
### PR changes

Others
### Describe

Add rebuild group feature in reducer.
In dynamic graph networking, if the parameter order is not declared in the forward order, performance may be affected. As shown in the figure: in the Bert model, putting the declaration of the embedding parameter of the first layer at the end will result in slower performance. The timeline is as follows:
image
Here add the automatic rebuild group strategy, in the second batch, the group ready order can be adjusted automatically, which makes the computation and communication stream overlap fully. From the experiment of 4 machines and 32 cards, the speed can be improved  about **10.6%**.

### CODE REVIEW

File: paddle/fluid/imperative/reducer.cc [hunk 1]
```diff
@@ -257,10 +275,31 @@ void Reducer::MarkGroupReady(size_t group_index) {
...
+  std::reverse(rebuild_group_index.begin(), rebuild_group_index.end());
+  return rebuild_group_index;
+}
+
 void Reducer::FinalizeBackward() {
   PADDLE_ENFORCE_CUDA_SUCCESS(cudaEventRecord(comm_enent_.get(), comm_stream_));
   PADDLE_ENFORCE_CUDA_SUCCESS(
       cudaStreamWaitEvent(compute_stream_, comm_enent_.get(), 0));
+  if (!has_rebuilt_group_) {
+    auto rebuild_group_index = RebuildGruops();
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc:296]: const auto&?

File: paddle/fluid/imperative/reducer.cc [hunk 2]
```diff
@@ -20,47 +20,98 @@ namespace imperative {
...
+      PADDLE_THROW(platform::errors::Unimplemented(
+          "Data type (%s) is not supported when it splits tensors for "
+          "allreduce.",
+          framework::DataTypeToString(dtype_)));
+  }
+}
+
+std::ostream &operator<<(std::ostream &out, const Group &group) {
+  const auto &vars = group.variable_indices_;
+  out << "numul: " << group.all_length_ << " ;is_sparse: " << group.is_sparse_
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc:71]: numul->numel

File: paddle/fluid/imperative/reducer.cc [hunk 3]
```diff
@@ -159,17 +223,27 @@ void Reducer::InitializeGroups(
...
           .mutable_data(place_, group.dtype_);
     }
-    // Debug Message For Reducer
-    VLOG(3) << "the groups_[" << group_index << "] basic message:";
-    VLOG(3) << "numul: " << all_length << " ;is_sparse: " << group.is_sparse_
-            << " ;var number: " << group.variable_indices_.size();
+
+    // map variables to this group by VariableLocator
+    size_t inside_group_index = 0;
+    for (const auto var_index : group_indices[group_index]) {
```
Comment (hutuxian) [inline paddle/fluid/imperative/reducer.cc:235]: use `variable_indices_` directly?

**Review by hutuxian — APPROVED** [review]
LGTM

File: paddle/fluid/imperative/reducer.cc (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc:71]: fix in next pr.

File: paddle/fluid/imperative/reducer.cc (hunk 3 above)
Comment (ForFishes) [reply paddle/fluid/imperative/reducer.cc:235]: fix in next pr.

**Pull request merged**